tt_eager/tt_dnn/op_library/topk/kernels/topk.cpp

```diff
@@ -0,0 +1,171 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+
+#include "compute_kernel_api.h"
+#include "compute_kernel_api/transpose_wh.h"
+#include "compute_kernel_api/tile_move_copy.h"
+#include "compute_kernel_api/unpack.h"
+#include "compute_kernel_api/pack.h"
+
+// topk llk needs a global variable atm
+// this can only be removed once that's fixed
+int32_t topk_replay_init = 0;
+
+namespace NAMESPACE {
+void MAIN {
+    constexpr uint32_t input_cb_index = get_compile_time_arg_val(0);
+    constexpr uint32_t index_cb_index = get_compile_time_arg_val(1);
+    constexpr uint32_t input_transposed_cb_index = get_compile_time_arg_val(2);
+    constexpr uint32_t index_transposed_cb_index = get_compile_time_arg_val(3);
+    constexpr uint32_t values_cb_index = get_compile_time_arg_val(4);
+    constexpr uint32_t output_ind_cb_index = get_compile_time_arg_val(5);
+    constexpr uint32_t Ht = get_compile_time_arg_val(6);
+    constexpr uint32_t Wt = get_compile_time_arg_val(7);
+    constexpr uint32_t K = get_compile_time_arg_val(8);
+    constexpr uint32_t logk = get_compile_time_arg_val(9);
+    constexpr uint32_t logWt = get_compile_time_arg_val(10);
+
+    // init pack, compute and unpack
+    transpose_wh_init(input_cb_index, input_transposed_cb_index);
+    for(uint32_t ht = 0; ht < Ht; ++ht) {
+        bool ascending = false;
+        cb_reserve_back(input_transposed_cb_index, Wt);
+        cb_reserve_back(index_transposed_cb_index, Wt);
+
+        for (uint32_t wt = 0; wt < Wt; wt+=2) {
+            acquire_dst(tt::DstMode::Half);
+            // local sort into k groups
+            cb_wait_front(input_cb_index, 2);
+            cb_wait_front(index_cb_index, 2);
+
+            unpack_reconfig_data_format_srca(input_cb_index);
+            transpose_wh_init_short(input_cb_index);
+            transpose_wh_tile(input_cb_index, wt, 0);
+            transpose_wh_tile(input_cb_index, wt+1, 1);
+
+            unpack_reconfig_data_format_srca(index_cb_index);
+            transpose_wh_init_short(index_cb_index);
+            transpose_wh_tile(index_cb_index, wt, 2);
+            transpose_wh_tile(index_cb_index, wt+1, 3);
+
+            // llk_topk_sort -> inplace
+            ckernel::topk_tile_init();
```

Comment:
This is the only sfpu kernel used in the op right? Does this work if we only use the init once outside the loop?

Reply:
Yeah it works, changed